src/Common.cpp

```diff
@@ -160,10 +149,30 @@ std::string NAS2D::toLowercase(const std::string& str)
 std::string NAS2D::toUppercase(const std::string& str)
 {
 	std::string transformStr(str);
-	std::transform(transformStr.begin(), transformStr.end(), transformStr.begin(), (int(*)(int))std::toupper);
+	std::transform(transformStr.begin(), transformStr.end(), transformStr.begin(), (int (*)(int))std::toupper);
 	return transformStr;
 }
 
+std::vector<std::string> NAS2D::split(std::string str, char delim /*= ','*/, bool skip_empty /*= true*/)
+{
+	const auto potential_count = 1 + std::count(std::begin(str), std::end(str), delim);
+	NAS2D::StringList result{};
+	result.reserve(potential_count);
+
+	std::stringstream ss{};
+	ss.str(str);
+	ss.seekg(0);
+	ss.seekp(0);
+	ss.clear();
```

Comment:
When setting the string value of a stringstream via `.str` it puts the seek pointers at the end of the stream and sets the bad state. The above method calls puts them back to the beginning and clears the bad state. This is particularly important because I immediately try to read from it using `std::getline`

Reply:
Ahh, interesting. I was just reading:
[`stringstream::str`](https://en.cppreference.com/w/cpp/io/basic_stringstream/str)
[`stringbuf::str`](https://en.cppreference.com/w/cpp/io/basic_stringbuf/str)

The [`seekp`](https://en.cppreference.com/w/cpp/io/basic_ostream/seekp) method sets the output "put" pointer. The [`seekg`](https://en.cppreference.com/w/cpp/io/basic_istream/seekg) sets the input "get" pointer. As you only seem to be reading from this stream, you probably only need to use the `seekg` variant.

Reading further, it seems the stream position after setting `str` is dependent on the `openmode` to the constructor. It may be possible to avoid having to set stream position by adjusting the `openmode`.

The `openmode` has a default set by the constructor. One way to adjust the `openmode`, is to change the underlying stream type. As you only seem to be reading from this stream, you can perhaps use an [`istringstream`](https://en.cppreference.com/w/cpp/io/basic_istringstream/basic_istringstream) rather than the bidirectional `stringstream`. Additionally, the constructor can take a `std::string` parameter, allowing you to collapse the `.str` call with object construction.

I'm also a little uncertain what conditions require [`clear()`](https://en.cppreference.com/w/cpp/io/basic_ios/clear) to need to be called here.

```cpp
std::istringstream ss(str);
ss.seekg(0); // Maybe? (This might depend on open mode)
ss.clear(); // Maybe? (This might also depend on open mode or on seeking)
```

Reply:
It's a bad habit of mine, trying to use `std::stringstream` everywhere because if I were to change the code to

    std::stringstream ss{};
    ss << str;
    //...
    while(std::getline(ss, /*...*/)) /*...*/

Then using either `istringstream` or `ostringstream` would not work. Additionally, when constructing-with-string, if you forget to append the open mode with `std::ios_base::ate` the seek positions do not change and subsequent `operator<<` calls will not put the string where intended.